### PR TITLE
PCC04 GeoffRiley

### PR DESCRIPTION
**Pull request by GeoffRiley**
Difficulty level (1-10): [4]
Estimated time spent (hours): [3]
Completed (yes/no): [Yes]
I stretched my coding skills (if yes what did you learn?): [Uncovering the back office of Twitters API has been quite interesting.  The only thing that really complicated matters was the rigmarole of hoops that Twitter present in getting API keys.]
Other feedback (what can we improve?): [The API key application system is different… apps.twitter.com no longer exists and you have to register as a developer before you can go on to 'create' an app to get keys to use.  I put it down to being useful for my Open University Masters project on databases, but I'm not sure what a none education person would subscribe as.
Oh, also I had to increase the version of `tweepy` to 3.7.0 before it would run with Python 3.7.]

### CODE REVIEW

**Comment by pybites** [comment]
Twitter API was easy at some time, shame they make it so difficult to get setup now :(

File: 04/GeoffRiley/usertweets.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    - save tweets to csv file in data/ subdirectory
+    - implement len() an getitem() magic (dunder) methods"""
+
+    def __init__(self, handle, max_id=None):
+        self._handle = handle
+        self._save_file = f'{DEST_DIR}/{handle}.{EXT}'
+        self._max_id = max_id
+        self._auth = tweepy.OAuthHandler(CONSUMER_KEY, CONSUMER_SECRET)
+        self._auth.set_access_token(ACCESS_TOKEN, ACCESS_SECRET)
+        self._api = tweepy.API(self._auth)
```
Comment (pybites) [inline 04/GeoffRiley/usertweets.py]: could make a helper

File: 04/GeoffRiley/usertweets.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        self.read_tweets()
+        self.save()
+
+    def read_tweets(self):
+        twitterings = self._api.user_timeline(screen_name=self._handle, max_id=self._max_id, count=NUM_TWEETS)
```
Comment (pybites) [inline 04/GeoffRiley/usertweets.py]: E501? not sure but I usually would use multiple lines

File: 04/GeoffRiley/usertweets.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        self._twitterings = [Tweet(x.id_str, x.created_at, x.text) for x in twitterings]
+
+    def save(self):
+        if not os.path.exists(DEST_DIR):
+            os.makedirs(DEST_DIR)
+        if os.path.exists(self._save_file):
+            os.remove(self._save_file)
+        with open(self._save_file, 'w') as f:
```
Comment (pybites) [inline 04/GeoffRiley/usertweets.py]: abstract into helper?

File: 04/GeoffRiley/usertweets.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+    @property
+    def output_file(self):
+        return self._save_file
+
+
+if __name__ == "__main__":
+
+    for handle in ('pybites', '_juliansequeira', 'bbelderbos'):
+        print('--- {} ---'.format(handle))
```
Comment (pybites) [inline 04/GeoffRiley/usertweets.py]: would be cool to make this a `__str__` dunder on the class I (Bob) think

**Comment by pybites** [comment]
Very nice!

File: 04/GeoffRiley/usertweets.py (hunk 2 above)
Comment (GeoffRiley) [reply 04/GeoffRiley/usertweets.py]: Actually, I don't know why I let such a long line get through… I usually go through and deal with them before I push to git.  I use PyCharm so it commonly flags up if I've missed any too,.  Odd.

I totally agree with you though.

File: 04/GeoffRiley/usertweets.py (hunk 1 above)
Comment (GeoffRiley) [reply 04/GeoffRiley/usertweets.py]: Yes, I understand that concept. My reason for not doing so was that as it's a self contained object it wasn't going to be changing any of those elements after the initialisation.

However, one development I did consider was to make the authorisation and connection as a separate singleton class in order to isolate that part of the activity… but it seemed a little overkill for this example :)

**Pull request merged**